radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/PendingOutboundChannelsManager.java

```diff
@@ -128,37 +128,43 @@ public PendingOutboundChannelsManager(
   }
 
   private void handlePeerConnected(PeerConnected peerConnected) {
+    final var channel = peerConnected.channel();
+    final Optional<CompletableFuture<PeerChannel>> channelFutureOpt;
     synchronized (lock) {
```

Comment:
Why do we need a lock here? Can't we use a ConcurrentMap?

Reply:
Possibly - sounds like a reasonable idea, it depends what the lock is protecting - is it just the thread safety of the map, or is the lock protecting other logical guarantees (eg about the number of connection attempts to a single node in connectTo).

On that note - @LukasGasior1  the lock in `connectTo` feels overly locked, and calls a `dispatch` inside the lock, which I don't like the look of, although dispatch is possibly thread-safe?  But I don't fully understand the implications of the locking here and what it's supposed to prevent, but if it can be reduced to just locking the Map, and so be replaced by a ConcurrentMap, that's not a bad idea :)

Reply:
@diegomrsantos we could use a ConcurrentMap and its `computeIfAbsent` method in `connectTo`, but as @dhedey pointed out, doing it with a manually managed lock allows us to extract `peerOutboundBootstrap.initOutboundConnection` and `timeoutEventDispatcher.dispatch` out of the lock in `connectTo` - which I did in the latest commit. With a ConcurrentMap these two calls would need to be in the `computeIfAbsent` callback, holding ConcurrentMap's lock.

Reply:
Not sure if I follow. It seems the only thing this lock is preventing is two concurrent removals from the map. In a concurrent map I would expect one to succeed and the other to return null and then nothing would be done.

Reply:
Just to clarify, I wasn't talking about the connectTo method, but the specific one where I added the comment. I was also talking about the use of the lock in this method not the lock in general. The other methods follow a similar pattern where the synchronize block is used only to protect concurrent removals from the map.

Reply:
In `connectTo` it synchronized the check whether the key is in the map, and the put operation, followed by some actions (in case the element was missing).
While the specific place where you added the comment could benefit from the ConcurrentMap, the `connectTo` method would still require an external lock (see my comment above) and I think it's better to stick to one approach rather than mixing ConcurrentMap's internal lock with an external object lock in a single class.

Reply:
The deadlock was caused by a lock that could be avoided and it seems the same can be said for all the other methods. What are the disadvantages of using the `ConcurrentMap` and manually locking only where there's no trivial alternative?

Reply:
Actually, reading it one more time, I can't see any difference between the `synchronised` block  in `connectTo` and using [putIfAbsent](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html#putIfAbsent-K-V-).

Reply:
Given my comment above, how would you avoid a lock for all other method? I don't quite see it. Would you mind submitting a commit for your suggested changes?
The disadvantage to me is that, because we still need to use the synchronized block elsewhere, there's no point of using a synchronized collection. In my opinion it's easier to reason about the code if only a single locking mechanism is used.
Additionally, if we used a concurrent collection, there would be some performance impact in `connectTo`, because of duplicated synchronization (internal in the collection and the outer synchronized block) [the impact is probably quite low, but in any case, it's one more argument on my side :laughing: ].

Reply:
Currently `peerOutboundBootstrap.initOutboundConnection` and `timeoutEventDispatcher.dispatch` are called after releasing the lock. With `putIfAbsent` we'd need to call these in the absent callback, while still holding the lock.
And we can't set a non-final flag variable from within the callback.

Reply:
hmh, although this could bo probably solved by using some mutable container, for example AtomicBoolean

Reply:
I submitted a commit. `putIfAbsent` is atomic. Please review if it makes sense. Is there a test that could be run?

Reply:
Thanks! I think it looks good, didn't realize that `putIfAbsent` returns the previous value **or null**.